### PR TITLE
Bulk Query Refactor

### PR DESCRIPTION
**Pull request by gregdaynes**
- Moved `GraphQL.BulkFetch` to a new top level namespace `Bulk`
- Added new option `auto_cancel`
- Refactored query functions - process! is cleaner
- Split execute query and fetch results into separate functions
- Cancel on query timeout can try to verify cancel

Co-authored-by: Hez <github@hez.ca>
Co-authored-by: Jason Freeman <pushpaw@gmail.com>
Co-authored-by: Mark Zsombor <markszsombor@gmail.com>

### CODE REVIEW

File: lib/shopify_api/bulk/query.ex [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    Process.sleep(polling_rate)
+
+    case status(token) do
+      {:ok, %{"status" => "COMPLETED", "url" => url} = _response} -> {:ok, url}
+      _ -> poll(token, bulk_query_id, polling_rate, max_poll, depth + 1)
+    end
+  end
+
+  defp maybe_cancel(false, _, _), do: nil
+  defp maybe_cancel(true, token, bid), do: token |> cancel(bid) |> poll_till_cancel(token, bid)
```
Comment (MarkZsombor) [inline lib/shopify_api/bulk/query.ex:176]: 🔥This shouldn't be passing `bid` as an argument to `poll_till_cancel` unless we want to start REALLY BIG max_poll values.

Comment (freemer) [reply lib/shopify_api/bulk/query.ex:176]: Hmm, the compile warning this causes is odd:
`the first 3 default arguments in poll_till_cancel/4 are never used`

When it's actually the first default argument, but the func's third arg.

File: lib/shopify_api/bulk/query.ex [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+  defp poll_till_cancel(resp, token, max_poll \\ 300, depth \\ 0)
+
+  defp poll_till_cancel({:ok, %{"bulkOperation" => %{"status" => "CANCELED"}}}, _token, _, _),
+    do: true
+
+  defp poll_till_cancel({:ok, %{"status" => "CANCELED"}}, _token, _, _), do: true
+
+  defp poll_till_cancel(_, _token, max_poll, depth) when max_poll == depth,
+    do: {:error, :cancelation_timedout}
```
Comment (MarkZsombor) [inline lib/shopify_api/bulk/query.ex:186]: We aren't doing anything with the return values for this function, and therefore not bubbling up a cancel timeout.

Comment (hez) [reply lib/shopify_api/bulk/query.ex:186]: Correct, do we want to differentiate between a cancel poll timeout and a general one?

File: .tool-versions [hunk 3]
```diff
@@ -1,2 +1,2 @@
-erlang 22.0.7
+erlang 22.2
```
Comment (hez) [inline .tool-versions:1]: accidental commit?

File: lib/shopify_api/bulk/query.ex [hunk 4]
```diff
@@ -177,11 +177,16 @@ defmodule ShopifyAPI.Bulk.Query do
...
 
+  # response from cancel
   defp poll_till_cancel({:ok, %{"bulkOperation" => %{"status" => "CANCELED"}}}, _token, _, _),
     do: true
 
+  # response from status
   defp poll_till_cancel({:ok, %{"status" => "CANCELED"}}, _token, _, _), do: true
 
+  # Sometimes operations complete before they are able to cancelled
+  defp poll_till_cancel({:ok, %{"status" => "COMPLETED"}}, _token, _, _), do: true
```
Comment (hez) [inline lib/shopify_api/bulk/query.ex:188]: nice! 👍

File: .tool-versions (hunk 3 above)
Comment (freemer) [reply .tool-versions:1]: Yep 🤦

File: lib/shopify_api/bulk/cancel.ex [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  require Logger
+
+  alias ShopifyAPI.AuthToken
+  alias ShopifyAPI.Bulk.Query
+
+  @polling_timeout_message "BulkFetch timed out before completion"
+  @auto_cancel_sleep_duration 1_000
+
+  @spec pretty_please(boolean(), AuthToken.t(), String.t()) :: {:ok | :error, any()}
+  # For historical reasons this joke function name must stay.
```
Comment (freemer) [inline lib/shopify_api/bulk/cancel.ex:11]: lol

File: lib/shopify_api/bulk/query.ex [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    |> HTTPoison.get()
+    |> case do
+      {:ok, %{body: jsonl}} -> {:ok, jsonl}
+      error -> error
+    end
+  end
+
+  def parse_response!(""), do: []
+  def parse_response!({:ok, jsonl}), do: parse_response!(jsonl)
+  def parse_response!({:error, msg}), do: raise(ShopifyAPI.Bulk.QueryError, msg)
```
Comment (MarkZsombor) [inline lib/shopify_api/bulk/query.ex:102]: We need to add another case here, for when msg is a struct we raise `inspect(msg)`

File: lib/shopify_api/bulk/cancel.ex (hunk 5 above)
Comment (MarkZsombor) [reply lib/shopify_api/bulk/cancel.ex:11]: This does need to be dealt with

**Comment by MarkZsombor** [comment]
This needs update to the changelog, as it is a breaking change to anyone else who is using bulk queries.

**Pull request merged**